internal/zinc-core/src/main/scala/sbt/internal/inc/IncrementalCommon.scala

```diff
@@ -11,321 +11,348 @@ package inc
     }
+  }
 
-  private[this] def recompileClasses(classes: Set[String],
-                                     modifiedSrcs: Set[File],
-                                     allSources: Set[File],
-                                     binaryChanges: DependencyChanges,
-                                     previous: Analysis,
-                                     doCompile: (Set[File], DependencyChanges) => Analysis,
-                                     classfileManager: XClassFileManager): (Analysis, Set[File]) = {
-    val invalidatedSources = classes.flatMap(previous.relations.definesClass) ++ modifiedSrcs
-    val invalidatedSourcesForCompilation = expand(invalidatedSources, allSources)
-    val pruned = Incremental.prune(invalidatedSourcesForCompilation, previous, classfileManager)
-    debug("********* Pruned: \n" + pruned.relations + "\n*********")
+  def mapInvalidationsToSources(
+      invalidatedClasses: Set[String],
+      aggregateSources: Set[File],
+      allSources: Set[File],
+      previous: Analysis
+  ): Set[File] = {
+    def expand(invalidated: Set[File]): Set[File] = {
+      val recompileAllFraction = options.recompileAllFraction
+      if (invalidated.size <= allSources.size * recompileAllFraction) invalidated
+      else {
+        log.debug(
+          s"Recompiling all sources: number of invalidated sources > ${recompileAllFraction * 100.00}% of all sources")
+        allSources ++ invalidated // Union because `all` doesn't contain removed sources
+      }
+    }
 
-    val fresh = doCompile(invalidatedSourcesForCompilation, binaryChanges)
-    // For javac as class files are added to classfileManager as they are generated, so
-    // this step is redundant. For scalac this is still necessary. TODO: do the same for scalac.
-    classfileManager.generated(fresh.relations.allProducts.toArray)
-    debug("********* Fresh: \n" + fresh.relations + "\n*********")
-    val merged = pruned ++ fresh //.copy(relations = pruned.relations ++ fresh.relations, apis = pruned.apis ++ fresh.apis)
-    debug("********* Merged: \n" + merged.relations + "\n*********")
-    (merged, invalidatedSourcesForCompilation)
+    expand(invalidatedClasses.flatMap(previous.relations.definesClass) ++ aggregateSources)
   }
 
-  private[this] def emptyChanges: DependencyChanges = new DependencyChanges {
-    val modifiedBinaries = new Array[File](0)
-    val modifiedClasses = new Array[String](0)
-    def isEmpty = true
-  }
-  private[this] def expand(invalidated: Set[File], all: Set[File]): Set[File] = {
-    val recompileAllFraction = options.recompileAllFraction
-    if (invalidated.size > all.size * recompileAllFraction) {
-      log.debug(
-        "Recompiling all " + all.size + " sources: invalidated sources (" + invalidated.size + ") exceeded " + (recompileAllFraction * 100.0) + "% of all sources")
-      all ++ invalidated // need the union because all doesn't contain removed sources
-    } else invalidated
-  }
+  def recompileClasses(
```

Comment:
Documentation, reorganization, and some changes are intermixed in this commit, so it's difficult for me to preview, but have you made any intentional change in the observable behavior, either in the difference in output or in performance characteristics? For example, I noticed that the method named `recompileClasses` are no longer recompiling classes, but it's recompiling sources.

Reply:
There’s no change in semantics, I double checked this several times manually and the tests confirm it.

Reply:
Wrt the change you mention @eed3si9n, the reason why you see that we recompile sources instead of classes in a clearer way is because `recompileClasses` was doing the mapping between invalidated classes and sources to recompile, and I pulled that logic out of it to keep things clear.